0002-crypto-conditions/0002-crypto-conditions.md

```diff
@@ -110,508 +115,638 @@ informative:
+  2. Calculate the maximum fulfillment length of a fulfillment
+  
+For compound types the fingerprint content will contain the complete, encoded, condition for all sub-crypto-conditions. Implementations MUST abide by the ordering rules provided when assembling the fingerprint content.
 
-Fulfillments are ASCII encoded as:
+When calculating the fingerprint of a compound crypto-condition implementations MUST first derive the condition for all sub-fulfillments and include these conditions when assembling the fingerprint content.
 
-    "cf:" BASE16(type) ":" BASE64URL(payload)
+## Conditions as Public Keys
 
-### Binary Format {#binary-fulfillment-format}
+Since the condition is just a fingerprint and meta-data about the crypto-condition it can be transmitted freely in the same way a public key is shared publicly. It's not possible to derive the fulfillment from the condition.
 
-Fulfillments are binary encoded as:
+# Format
 
-    Fulfillment ::= SEQUENCE {
-      type ConditionType,
-      payload OCTET STRING
-    }
+A description of crypto-conditions is provided in this document using Abstract Syntax Notation One (ASN.1) as defined in [ITU.X680.2015](#itu.X680.2015).
 
-### Fields {#fulfillment-format-fields}
+## Encoding Rules
 
-type
-: is the numeric type identifier representing the condition type. For some condition types the fulfillment will contain further subfulfillments, however the type field always represents the outermost, or top-level, type.
+Implementations of this specificiation MUST support encoding and decoding using Distinguished Encoding Rules (DER) as defined in [ITU.X690.2015](#itu.X690.2015). This is the canonical encoding format.
 
-payload
-: The payload is an octet string whose internal format is defined by each of the types.
+Alternative encodings may be used to represent top-level conditions and fulfillments but to ensure a determinisitic outcome in producing the condition fingerprint content, including any sub-conditions, MUST be DER encoded prior to hashing (if hashing is required).
 
-### Example Fulfillment
+## Condition {#condition-format}
 
-The following is an example fulfillment in string format, for the [example condition](#example-condition):
+The binary encoding of conditions differs based on their type. All types define at least a fingerprint and maxFulfillmentLength sub-field. Some types, such as the compound condition types, define additional sub-fields that are required to convey essential properties of the crypto-condition (such as the sub-types used by sub-conditions in the case of the compound types).
 
-    cf:0:VGhlIG9ubHkgYmFzaXMgZm9yIGdvb2QgU29jaWV0eSBpcyB1bmxpbWl0ZWQgY3JlZGl0LuKAlE9zY2FyIFdpbGRl
+Each crypto-condition type has a type ID. The list of known types is the IANA-maintained [Crypto-Condition Type Registry](#crypto-conditions-type-registry).
 
+Conditions are encoded as follows:
 
-The example has the following attributes:
+    Condition ::= CHOICE {
+	  preimageSha256 SimpleCondition,
```

Comment:
The tags are assumed to be the index if not provided.
I can add them but it won't affect the encoding.